include/envoy/buffer/buffer.h

```diff
@@ -117,6 +117,31 @@ class Instance {
    */
   virtual void addBufferFragment(BufferFragment& fragment) PURE;
 
+  /**
+   * Insert externally owned data into the buffer after given RawSlice and its index. No copying is
+   * done. fragment is not owned. When the fragment->data() is no longer needed, fragment->done() is
+   * called. If RawSlice.mem_ and RawSlice.len_ not same in the Instance, BufferFragment insertion
+   * is not performed and false is returned.
+   *
+   * @param i index of RawSlice in RawSlice array obtained from getRawSlices method.
+   * @param slice slice from RawSlice array obtained from getRawSlices method.
+   * @param fragment the externally owned data to insert in to the buffer.
+   * @return true if insertion was successful or false and no change is made to the buffer
+   */
+  virtual bool insertBufferFragmentAfter(size_t i, RawSlice& slice, BufferFragment& fragment) PURE;
```

Comment:
Most STL containers that do support insert usually just follow a convention of inserting before given element X, could we possibly do that here as well to narrow the interface? Maybe supporting a stronger iterator is the way to go (i.e. a bidirecitonal iterator)

Reply:
Added Bidirectional iterator to buffer. Yet to implement insert that takes and iterator and input.